tool/teleport-update/main.go

```diff
@@ -0,0 +1,251 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package main
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"log/slog"
+	"net/http"
+	"net/url"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"syscall"
+	"time"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/net/http/httpproxy"
+
+	"github.com/gravitational/teleport"
+	apidefaults "github.com/gravitational/teleport/api/defaults"
+	tracehttp "github.com/gravitational/teleport/api/observability/tracing/http"
+	apiutils "github.com/gravitational/teleport/api/utils"
+	"github.com/gravitational/teleport/lib/autoupdate"
+	libdefaults "github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/modules"
+	libutils "github.com/gravitational/teleport/lib/utils"
+	logutils "github.com/gravitational/teleport/lib/utils/log"
+)
+
+const appHelp = `Teleport Updater
+
+The Teleport Updater updates the version a Teleport agent on a Linux server
+that is being used as agent to provide connectivity to Teleport resources.
+
+The Teleport Updater supports upgrade schedules and automated rollbacks. 
+
+Find out more at https://goteleport.com/docs/updater`
+
+const (
+	// templateEnvVar allows the template for the Teleport tgz to be specified via env var.
+	templateEnvVar = "TELEPORT_URL_TEMPLATE"
+	// proxyServerEnvVar allows the proxy server address to be specified via env var.
+	proxyServerEnvVar = "TELEPORT_PROXY"
+	// updateGroupEnvVar allows the update group to be specified via env var.
+	updateGroupEnvVar = "TELEPORT_UPDATE_GROUP"
+)
+
+const (
+	// versionsDirName specifies the name of the subdirectory inside of the Teleport data dir for storing Teleport versions.
+	versionsDirName = "versions"
+	// configFileName specifies the name of the file inside versionsDirName containing configuration for the teleport update
+	configFileName = "update.yaml"
+	// lockFileName specifies the name of the file inside versionsDirName containing the flock lock preventing concurrent updater execution.
+	lockFileName = ".lock"
+)
+
+var plog = logutils.NewPackageLogger(teleport.ComponentKey, teleport.ComponentUpdater)
```

Comment:
Can we pass loggers along rather than use package-level loggers? It's going to be annoying if/when we decide we want to embed some of lib/autoupdate in Teleport and it becomes yet another part that doesn't behave properly with regards to logging.